### PR TITLE
linting - jsx a11y errors

### LINKED ISSUE #2563: Linting - jsx-a11y
Fix all instances of the linting errors associated with jsx-ally 
If necessary, add a rule disable comment for instances where appropriate and add a justification comment.

Tips:

- Run npm run lint --workspaces from the frontend folder to manually run linting on all subprojects
- The [eslint-formatter-summary](https://github.com/mhipszki/eslint-formatter-summary) package can be handy for generating a linting summary by rule

### PR DESCRIPTION
**Pull request by vd1992**
resolves #2563 

Think I got all them and fixed without things getting funky. One I'm not sure is properly fixable, it wants the label to nest around `<inputs>` and I think the input label being structured as it is makes it throw errors. To fix it might require refactoring the inputs group of components?

### CODE REVIEW

**Comment by vd1992** [comment]
I can see the failing test, I don't understand it though

**Comment by vd1992** [comment]
I can see the failing test, I don't understand it though

**Comment by petertgiles** [comment]
> I don't understand it though

Suggested steps:
1) Run test suite locally, verify you get the same failure
2) Figure out which test is failing and run just that test
3) Study the test to understand what it's trying to do
4) I think Jest automatically dumps the dom out to the console when a test is failed - take a look and see why it's failing

**Comment by vd1992** [comment]
@petertgiles 
okay, I think it is good now

File: frontend/common/src/components/inputPartials/Fieldset/Fieldset.stories.tsx [hunk 1]
```diff
@@ -11,17 +11,17 @@ const TemplateFieldset: Story<FieldsetProps> = (args) => {
   const { name } = args;
   return (
     <Fieldset {...args}>
-      <label data-h2-display="b(block)">
+      <label data-h2-display="b(block)" htmlFor={name}>
         One
-        <input type="checkbox" name={name} value="One" />
+        <input type="checkbox" name={name} value="One" id={name} />
```
Comment (petertgiles) [inline frontend/common/src/components/inputPartials/Fieldset/Fieldset.stories.tsx:16]: Shouldn't these id values be unique in the fieldset?

File: frontend/common/src/components/form/Checkbox/Checkbox.tsx [hunk 2]
```diff
@@ -79,7 +79,7 @@ export const Checkbox: React.FunctionComponent<CheckboxProps> = ({
             data-h2-align-items="b(center)"
             style={{ width: "100%" }}
           >
-            <label>
+            <label htmlFor={name}>
```
Comment (petertgiles) [inline frontend/common/src/components/form/Checkbox/Checkbox.tsx:82]: Shouldn't this be `id`?

File: frontend/common/src/components/inputPartials/InputLabel/InputLabel.tsx [hunk 3]
```diff
@@ -1,3 +1,5 @@
+/* eslint-disable jsx-a11y/label-has-for */
```
Comment (petertgiles) [inline frontend/common/src/components/inputPartials/InputLabel/InputLabel.tsx:1]: If you want to zap out the entire file, please add a new issue to the board and note the issue number here so we can come back and give it another try sometime.

File: frontend/talentsearch/src/js/components/skills/SkillChecklist/SkillChecklist.tsx [hunk 4]
```diff
@@ -30,15 +38,15 @@ const Family: React.FunctionComponent<FamilyProps> = ({
...
-        onChange={(e) => {
-          callback(family, e.target.checked);
-        }}
-      />
-      &nbsp;
-      <label htmlFor={family.id}>
+      <label htmlFor={htmlForAssigner}>
+        <input
+          type="checkbox"
+          id={htmlForAssigner}
```
Comment (petertgiles) [inline frontend/talentsearch/src/js/components/skills/SkillChecklist/SkillChecklist.tsx:44]: I'm not clear on why this change is necessary.  Can't the id stay as `family.id`?

**Review by petertgiles — CHANGES_REQUESTED** [review]
Cool stuff!  I'll admit I'm a little out of my depth here and I'm sure that there are people on the team who are much more knowledgeable in this domain.  Anyway, some of my thoughts below.

File: frontend/talentsearch/src/js/components/skills/SkillChecklist/SkillChecklist.tsx (hunk 4 above)
Comment (vd1992) [reply frontend/talentsearch/src/js/components/skills/SkillChecklist/SkillChecklist.tsx:44]: jest was throwing errors about "absence of a `for`" this fixed it
I think it wanted labels and id to match

File: frontend/common/src/components/form/Checkbox/Checkbox.tsx (hunk 2 above)
Comment (vd1992) [reply frontend/common/src/components/form/Checkbox/Checkbox.tsx:82]: I think either could work, let us ask Eric
@esizer 
Any thoughts on this, which it should be?

Comment (esizer) [reply frontend/common/src/components/form/Checkbox/Checkbox.tsx:82]: Yeah, the `htmlFor` attribute should reference the `id` of the associated input and must be unique as well.

File: frontend/common/src/components/inputPartials/InputLabel/InputLabel.tsx (hunk 3 above)
Comment (vd1992) [reply frontend/common/src/components/inputPartials/InputLabel/InputLabel.tsx:1]: @esizer 
I think linting was upset because `label` couldn't be conveniently wrapped around `input`
From which I thought this would be a complex fix with needed refactoring, do you want to weigh in on whether it would be a complex fix or not? Feel free to check out the latest commit if you want to poke around

File: frontend/talentsearch/src/js/components/skills/SkillChecklist/SkillChecklist.tsx (hunk 4 above)
Comment (petertgiles) [reply frontend/talentsearch/src/js/components/skills/SkillChecklist/SkillChecklist.tsx:44]: Just using the label wrapping that you did with a plain family.id seems to be OK for me.  Are we getting different errors again?
image

File: frontend/common/src/components/inputPartials/InputLabel/InputLabel.tsx (hunk 3 above)
Comment (esizer) [reply frontend/common/src/components/inputPartials/InputLabel/InputLabel.tsx:1]: Yeah, so this is a strange thing that most linters cannot handle. Since the label and input are in different files the linter cannot determine if they are associated with one another using `htmlFor`.

It will take a bit of extra work, but you can make the rule aware of these components.

Ref: https://github.com/jsx-eslint/eslint-plugin-jsx-a11y/blob/main/docs/rules/label-has-associated-control.md

```json
{
  "rules": {
    "jsx-a11y/label-has-associated-control": [ 2, {
      "labelComponents": ["CustomInputLabel"],
      "labelAttributes": ["label"],
      "controlComponents": ["CustomInput"],
      "depth": 3,
    }],
  }
}
```

File: frontend/talentsearch/src/js/components/skills/SkillChecklist/SkillChecklist.tsx (hunk 4 above)
Comment (esizer) [reply frontend/talentsearch/src/js/components/skills/SkillChecklist/SkillChecklist.tsx:44]: I'd be curious to see the error since nesting an `input` within a `label` creates an association so the `htmlFor` attribute should not be needed.

**Review by esizer — CHANGES_REQUESTED** [review]
I think we should take a look [at the rule itself](https://github.com/jsx-eslint/eslint-plugin-jsx-a11y/blob/main/docs/rules/label-has-associated-control.md) and see about tweaking it to meet our needs.

- [ ] Add custom components to the rule
- [ ] Ensure we are allowing nesting as a valid association

File: frontend/talentsearch/src/js/components/skills/SkillChecklist/SkillChecklist.tsx (hunk 4 above)
Comment (vd1992) [reply frontend/talentsearch/src/js/components/skills/SkillChecklist/SkillChecklist.tsx:44]: @esizer 
If I remove the `htmlFor`

image

Comment (vd1992) [reply frontend/talentsearch/src/js/components/skills/SkillChecklist/SkillChecklist.tsx:44]: okey, done in the last commit

**Comment by esizer** [comment]
> > I think we should take a look [at the rule itself](https://github.com/jsx-eslint/eslint-plugin-jsx-a11y/blob/main/docs/rules/label-has-associated-control.md) and see about tweaking it to meet our needs.
> > ```
> > * [ ]  Add custom components to the rule
> > 
> > * [ ]  Ensure we are allowing nesting as a valid association
> > ```
> 
> I don't think I know how to do that

Sorry, I made mention to this in [another comment](https://github.com/GCTC-NTGC/gc-digital-talent/pull/2575#discussion_r861777068) and really should have made a single comment to provide context.

```tsx
{
  "rules": {
    "jsx-a11y/label-has-associated-control": [ 2, {
      "labelComponents": ["CustomInputLabel"],
      "labelAttributes": ["label"],
      "controlComponents": ["CustomInput"],
      "depth": 3,
    }],
  }
}
```

We would need to edit the `.eslintrc` to add the `labelComponents` and `controlComponents` properties to the config.

Essentially, you would want to include the names of the components in the `labelComponents` and `controlComponents` arrays where the `controlComponents` are a list of the custom input components. In the example, `labelAttributes` are the props passed to the label component that represent the accessible name (if it just the children of the component, this is not necessary).

Hopefully that makes sense. If it doesn't, feel free to ping me with any questions or we could do a call to walk through it.

File: frontend/talentsearch/src/js/components/skills/SkillChecklist/SkillChecklist.tsx (hunk 4 above)
Comment (esizer) [reply frontend/talentsearch/src/js/components/skills/SkillChecklist/SkillChecklist.tsx:44]: Nice! 👍

**Comment by vd1992** [comment]
> > > I think we should take a look [at the rule itself](https://github.com/jsx-eslint/eslint-plugin-jsx-a11y/blob/main/docs/rules/label-has-associated-control.md) and see about tweaking it to meet our needs.
> > > ```
> > > * [ ]  Add custom components to the rule
> > > 
> > > * [ ]  Ensure we are allowing nesting as a valid association
> > > ```
> > 
> > 
> > I don't think I know how to do that
> 
> Sorry, I made mention to this in [another comment](https://github.com/GCTC-NTGC/gc-digital-talent/pull/2575#discussion_r861777068) and really should have made a single comment to provide context.
> 
> ```tsx
> {
>   "rules": {
>     "jsx-a11y/label-has-associated-control": [ 2, {
>       "labelComponents": ["CustomInputLabel"],
>       "labelAttributes": ["label"],
>       "controlComponents": ["CustomInput"],
>       "depth": 3,
>     }],
>   }
> }
> ```
> 
> We would need to edit the `.eslintrc` to add the `labelComponents` and `controlComponents` properties to the config.
> 
> Essentially, you would want to include the names of the components in the `labelComponents` and `controlComponents` arrays where the `controlComponents` are a list of the custom input components. In the example, `labelAttributes` are the props passed to the label component that represent the accessible name (if it just the children of the component, this is not necessary).
> 
> Hopefully that makes sense. If it doesn't, feel free to ping me with any questions or we could do a call to walk through it.

Okay, I think I got what you meant
I made changes in all the .eslintrc files in the four directories

**Comment by esizer** [comment]
> > > > I think we should take a look [at the rule itself](https://github.com/jsx-eslint/eslint-plugin-jsx-a11y/blob/main/docs/rules/label-has-associated-control.md) and see about tweaking it to meet our needs.
> > > > ```
> > > > * [ ]  Add custom components to the rule
> > > > 
> > > > * [ ]  Ensure we are allowing nesting as a valid association
> > > > ```
> > > 
> > > 
> > > I don't think I know how to do that
> > 
> > 
> > Sorry, I made mention to this in [another comment](https://github.com/GCTC-NTGC/gc-digital-talent/pull/2575#discussion_r861777068) and really should have made a single comment to provide context.
> > ```tsx
> > {
> >   "rules": {
> >     "jsx-a11y/label-has-associated-control": [ 2, {
> >       "labelComponents": ["CustomInputLabel"],
> >       "labelAttributes": ["label"],
> >       "controlComponents": ["CustomInput"],
> >       "depth": 3,
> >     }],
> >   }
> > }
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > We would need to edit the `.eslintrc` to add the `labelComponents` and `controlComponents` properties to the config.
> > Essentially, you would want to include the names of the components in the `labelComponents` and `controlComponents` arrays where the `controlComponents` are a list of the custom input components. In the example, `labelAttributes` are the props passed to the label component that represent the accessible name (if it just the children of the component, this is not necessary).
> > Hopefully that makes sense. If it doesn't, feel free to ping me with any questions or we could do a call to walk through it.
> 
> Okay, I think I got what you meant I made changes in all the .eslintrc files in the four directories

Cool, I just pulled it down and I think we cleared out all of the errors! 👍

**Review by esizer — APPROVED** [review]

File: frontend/talentsearch/src/js/components/skills/SkillChecklist/SkillChecklist.tsx (hunk 4 above)
Comment (esizer) [reply frontend/talentsearch/src/js/components/skills/SkillChecklist/SkillChecklist.tsx:44]: I could be wrong here but I think we should be able to remove the `id` completely since it seems it was only being used to satisfy the deprecated `jsx-a11y/label-has-for` rule.

Comment (vd1992) [reply frontend/talentsearch/src/js/components/skills/SkillChecklist/SkillChecklist.tsx]: @esizer @petertgiles 
`id` removed

**Review by petertgiles — APPROVED** [review]
